store/src/main/java/org/apache/rocketmq/store/DefaultMessageStore.java

```diff
@@ -993,6 +985,29 @@ public QueryMessageResult queryMessage(String topic, String key, int maxNum, lon
         return queryMessageResult;
     }
 
+    /**
+     * check target msg match
+     *
+     * @param msg   message
+     * @param topic topic
+     * @param key   message key
+     * @return  is match
+     */
+    private boolean isMessageMatch(MessageExt msg, String topic, String key) {
+        String keys = msg.getKeys();
+        if (Objects.equals(topic, msg.getTopic()) && StringUtils.isNotEmpty(keys)) {
```

Comment:
Filtering here may reduce the number of messages. It is recommended to filter in the selectPhyOffset() method.

Reply:
filter in selectPhyOffset() may increase read commitlog times, if this is acceptable, I'll modify it

Reply:
IMO, it's correct to filtering here, as we only have physical offsets in `selectPhyOffset()`, so we can't check the topic and key of message in `selectPhyOffset()`

Reply:
Yes, I agree with you.

Reply:
can we put check method in class `DefaultMessageStore` like this ? :

```
QueryOffsetResult queryOffsetResult = this.indexService.queryOffset(topic, key, maxNum, begin, lastQueryMsgTime, physicOffset -> {
    SelectMappedBufferResult result = commitLog.getData(physicOffset, false);
    return isTopicAndKeyMatch(result, topic, key);
});
```

if so, we will do more once IO and protocol analysis